extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java

```diff
@@ -119,233 +103,105 @@ public void mongoClientNames(ApplicationArchivesBuildItem applicationArchivesBui
+                        bsonDiscriminator.getBsonDiscriminatorClassNames(),
+                        poolListenerList, sslNativeConfig.isExplicitlyDisabled()))
+                .done());
 
-        return new BeanContainerListenerBuildItem(recorder.addMongoClient(
-                (Class<? extends AbstractMongoClientProducer>) recorderContext.classProxy(mongoClientProducerClassName),
-                sslNativeConfig.isExplicitlyDisabled()));
+        mongoConnections.produce(new MongoConnectionNameBuildItem(MongoClientBeanUtil.DEFAULT_MONGOCLIENT_NAME));
+        for (MongoClientNameBuildItem bi : mongoClientNames) {
+            mongoConnections.produce(new MongoConnectionNameBuildItem(bi.getName()));
+        }
     }
 
-    @Record(RUNTIME_INIT)
+    @Record(ExecutionTime.RUNTIME_INIT)
     @BuildStep
-    void configureRuntimePropertiesAndBuildClients(MongoClientRecorder recorder,
-            CodecProviderBuildItem codecProvider, BsonDiscriminatorBuildItem bsonDiscriminator,
-            List<MongoConnectionPoolListenerBuildItem> connectionPoolListenerProvider,
+    void generateClientBeans(MongoClientRecorder recorder,
             List<MongoClientNameBuildItem> mongoClientNames,
-            MongodbConfig mongodbConfig, ConfigurationBuildItem config,
-            BuildProducer<MongoConnectionNameBuildItem> mongoConnections) {
+            MongodbConfig mongodbConfig,
+            Optional<MongoUnremovableClientsBuildItem> mongoUnremovableClientsBuildItem,
+            BuildProducer<SyntheticBeanBuildItem> syntheticBeanBuildItemBuildProducer) {
+
+        boolean makeUnremovable = mongoUnremovableClientsBuildItem.isPresent();
+
+        // default blocking client
+        syntheticBeanBuildItemBuildProducer.produce(createSyntheticBean(recorder, mongodbConfig, makeUnremovable,
+                MongoClientBeanUtil.DEFAULT_MONGOCLIENT_NAME, false));
+        // default reactive client
+        syntheticBeanBuildItemBuildProducer.produce(createSyntheticBean(recorder, mongodbConfig, makeUnremovable,
+                MongoClientBeanUtil.DEFAULT_MONGOCLIENT_NAME, true));
+
+        for (MongoClientNameBuildItem mongoClientName : mongoClientNames) {
+            // named blocking client
+            syntheticBeanBuildItemBuildProducer
+                    .produce(createSyntheticBean(recorder, mongodbConfig, makeUnremovable, mongoClientName.getName(), false));
+            // named reactive client
+            syntheticBeanBuildItemBuildProducer
+                    .produce(createSyntheticBean(recorder, mongodbConfig, makeUnremovable, mongoClientName.getName(), true));
+        }
+    }
 
-        List<ConnectionPoolListener> poolListenerList = connectionPoolListenerProvider.stream()
-                .map(MongoConnectionPoolListenerBuildItem::getConnectionPoolListener)
-                .collect(Collectors.toList());
+    private SyntheticBeanBuildItem createSyntheticBean(MongoClientRecorder recorder, MongodbConfig mongodbConfig,
```

Comment:
I don't like the boolean to switch from the sync and reactive client. I found it not really readable.
Can we have two methods instead ?

Reply:
Sure, I was on the fence about this myself.

Reply:
Fixed